Shift without padding

naoyam opened the pull request:
Adds a boolean option to `shift` to disable padding. The primary motivation is to support
 the hdiff stencil example, which should now be correctly supported. 

Shifting without padding only sets values in a range that correspond to the valid range
in the input tensor. For example, suppose a tensor is defined from 0 to
N, shifting it to right by one produces an output that is defined only
from 1 to N. Similarly, if it is shifted to left, the output is defined
from 0 to N-1. Only valid range is adjusted, and the overall extent is
kept. IterDomain::start_ and IterDomain::stop_ define the valid range of
a domain.

Valid range is only defined in root domains. When a "partial" domain is
merged or split, the output domain does not inherit the range
information directly as it is not possible to define the valid range
just using two values. As such, loops still start from 0 to
IterDomain::extent. Predicates are used to enforce the valid range
defined in root domains.

It was necessary to modify blockReduce (and blockWelford) to support
non-zero start values since they gather the final result to thread 0
even when it might be disabled to run by read_write_pred.  When start is
not zero, the final output, which is held by thread 0, is not written to
the output tensor. This is fixed by having two predicates: one for reads
and another for writes. The read predicate is the same as before, but
the write predicate only looks at non-reducation axes so that the final
output is properly stored by thread 0. This separation of predicates is
only done when reduction axes have non-zero start values.

There's one remaining issue with syncthreads insertion. The WAR sync
insertion assumes each loop body is always executed at least once, which
may not be a valid assumption when the start value is not zero. Since
the only way non-zero start is introduced is currently through shifting,
and it's highly unlikely that start becomes as large as stop, so it's
not an immediate concern. I'll address it in a separate follow-up PR.

Update: The WAR sync issue is addressed in #1043

Commit:
Shift without padding

Adds a boolean option to `shift` to disable padding. Shifting without
padding only sets values in a range that correspond to the valid range
in the input tensor. For example, suppose a tensor is defined from 0 to
N, shifting it to right by one produces an output that is defined only
from 1 to N. Similarly, if it is shifted to left, the output is defined
from 0 to N-1. Only valid range is adjusted, and the overall extent is
kept. IterDomain::start_ and IterDomain::stop_ define the valid range of
a domain.

Valid range is only defined in root domains. When a "partial" domain is
merged or split, the output domain does not inherit the range
information directly as it is not possible to define the valid range
just using two values. As such, loops still start from 0 to
IterDomain::extent. Predicates are used to enforce the valid range
defined in root domains.

It was necessary to modify blockReduce (and blockWelford) to support
non-zero start values since they gather the final result to thread 0
even when it might be disabled to run by read_write_pred.  When start is
not zero, the final output, which is held by thread 0, is not written to
the output tensor. This is fixed by having two predicates: one for reads
and another for writes. The read predicate is the same as before, but
the write predicate only looks at non-reducation axes so that the final
output is properly stored by thread 0. This separation of predicates is
only done when reduction axes have non-zero start values.

There's one remaining issue with syncthreads insertion. The WAR sync
insertion assumes each loop body is always executed at least once, which
may not be a valid assumption when the start value is not zero. Since
the only way non-zero start is introduced is currently through shifting,
and it's highly unlikely that start becomes as large as stop, so it's
not an immediate concern. I'll address it in a separate follow-up PR.

Commit:
cleanup

Commit:
clang-tidy

Commit:
Remove empty IR nodes

CudaKernelGenerator attempts to simplify generated code with
statically predicated IfThenElse blocks. However, with a kernel like
below, it still ends up generating an empty else block.

```
if (...) {
  do_something
} else {
  if (false) {
  }
}
```

Generated code would look like:

```
if (...) {
  do_something
} else {
}
```

WIth `KIRCleaner::cleanUp`, the generated code would just look like:

```
if (...) {
  do_something
}
```

Commit:
cleanup

Commit:
Merge branch '20_12_3_devel' into shift-no-padding

csarofeen commented on torch/csrc/jit/codegen/cuda/arith.cpp:
```diff
@@ -1175,16 +1274,87 @@ TensorView* sum_to(TensorView* in, const std::vector<int64_t>& sum_to_size) {
+
+      if (offset > 0) {
+        // shift to right; extent remains the same, start and stop
+        // positions are moved right
```
Stop position seems to be getting `- offset`, so isn't it moving left?

naoyam replied:
Positive offsets move tensors right. Maybe "offsets" isn't a good name. It specifies a direction to shift.

The reason that the stop position is also adjusted is because it may not be equal to the extent. Since this is to shift to right, the stop position is also shifted right, i.e., `stop + offset`, and the `max` expression is because since `stop` never gets larger than `extent`. These adjustment computations are done statically when possible, but if not, it uses Fusion expressions to represent the new stop value.

csarofeen commented on torch/csrc/jit/codegen/cuda/index_compute.cpp:
```diff
@@ -2246,38 +2265,57 @@ Index::getReferenceRootPredicates(
+    kir::Val* stop = kir_contig_id->extent();
 
+    // TODO: This isn't used for now. When the consumer has halo,
+    // ShiftPredicateInserter is used.
```
Is the goal to use this instead of shift predicate inserter?

naoyam replied:
Yes, I think that'd be better. Originally, the shift predicate inserter wasn't that much different from the normal path. It was a sort of a thin extension that adds lower-end predicates. However, since then, the normal path is extended with the reference-based approach, and the shift predicate hasn't been updated yet. I don't see any fundamental difficulty to merge them.

csarofeen commented on torch/csrc/jit/codegen/cuda/ir_internal_nodes.h:
```diff
@@ -495,11 +509,18 @@ class TORCH_CUDA_CU_API IterDomain : public Val {
     return extent_;
   }
 
+  //! True if the valid range may be partial
```
Could you change the comment to something like `True if range of iteration domain isn't across the full extent."

csarofeen commented on torch/csrc/jit/codegen/cuda/lower2device.cpp:
```diff
@@ -169,7 +169,107 @@ std::unordered_map<Val*, Val*> getSimplificationMap(Fusion* fusion) {
+  }
+
+  void visit(kir::ForLoop* fl) final {
+    auto exprs = fl->body().exprs();
```
This function seems off to me. I would have expected `is_nop_` to be scoped.
In other words:
1) Save is_nop_
2) Reset is_nop_ to true
3) Traverse body
4) If is_nop_ is still true, clear body
5) reset is_nop_ to value from (1) & (4)

I would have thought the ite traversal would be like this too.

naoyam replied:
That was actually how I originally did, but I realized scoping isn't really necessary. `is_nop_` is just like a return value from visiting a child node, and if it is true, it's immediately removed (at line 205). If the loop body becomes empty, it sets the flag to true, i.e., signaling its parent node that this loop is empty, so at the parent node, the loop node is removed.

The removal transformation is done bottom up. For scoped nodes like `ForLoop` and `IfThenElse`, if their expr bodies are empty, they set `is_nop_` to true, and then removed by their parent nodes.

The `IfThenElse` handle didn't actually do this, which is fixed in 5e970ec42e.

csarofeen commented on torch/csrc/jit/codegen/cuda/lower_predicate.cpp:
```diff
@@ -48,6 +48,25 @@ class ConditionalFromPredicateModifier {
+      } else {
+        // If generateConditional returns null, use the same
+        // predicate as reads
+        expr->writePredicate()->setValue(read_cond);
```
Can't you just leave it as nullptr and get rid of the extra equivalency check in codegen.cpp?

naoyam replied:
Thanks. Updated.

csarofeen reviewed (changes requested):
Did you check all instances of `new IterDomain` to make sure stop is propagated where necessary?

In the future I believe we will want:
1) padding that's also an integer, not just true or false.
2) Padded value that can be set.
Maybe both of these are only relevant for gather?

Overall PR looks good to me. Only comment I want addressed before I can mark approved is the lower2device.cpp one.

Commit:
Separate read and write predicates for block and grid reductions

For block reductions, it's necessary when reduction axes may not start
with zero. For grid reductions, see issue #1049.

Closes #1049

Commit:
bug fix

Commit:
Additional tests

Commit:
Merge branch '20_12_3_devel' into shift-no-padding

Commit:
Disallow rfactor with partial domains since the valid range of a partial domain is only defined at its root domain.

Commit:
clang-format

Commit:
Merge branch '20_12_3_devel' into shift-no-padding

Commit:
PR feedback

naoyam commented:
> Did you check all instances of new IterDomain to make sure stop is propagated where necessary?

There was one place where I forgot to do so, which is the rfactor transformation. However, as we discussed, rfactor with partial domains isn't something we can support, so I added a validation assert there.

Commit:
bug fix

naoyam commented:
> In the future I believe we will want:
>
> padding that's also an integer, not just true or false.
> Padded value that can be set.
> Maybe both of these are only relevant for gather?

👍 

I'm also thinking about adding a new option like "assume-padded", which just assumes tensors are properly padded, which is actually how the Hdiff stencil is implemented.

Commit:
Fixing a bug introduced by the previous "fix" commit

Commit:
Merge branch '20_12_3_devel' into shift-no-padding

Commit:
Merge branch '20_12_3_devel' into shift-no-padding

csarofeen commented on test/cpp/jit/test_gpu_shift.cpp:
```diff
@@ -2859,12 +2866,316 @@ TEST(NVFuserTest, FusionConv2DStaticEvenSizedWindow_CUDA) {
+  tv1->split(0, 4);
+  tv1->split(-1, 8);
+  tv1->reorder({{1, 2}});
+  tv1->merge(-2, -1);
```
Could you change one of these to use transform propagator and parallelize all like? Can be in a follow up PR.

naoyam replied:
#1092

csarofeen commented on torch/csrc/jit/codegen/cuda/lower_shift.cpp:
```diff
@@ -293,14 +303,14 @@ kir::Bool* ShiftPredicateInserter::getPredicate(
                 ->as<kir::Bool>();
       }
     } else {
```
This else is pretty far down, can you just comment its if not shift expr.

csarofeen reviewed (approved):
Two minor comments that can be addressed any time, merging.
